### PR TITLE
Mqtt5Client Clean Shutdown Refactor

### PR DESCRIPTION
**Pull request by xiazhvera**
*Issue #, if available:*

*Description of changes:*
Added a Mqtt5ClientCore, and the client core will handle all native client callback related features.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by TwistedTwigleg — APPROVED** [review]
Looks good! There's a few things I left questions/comments on, but nothing major at all.

It might not hurt to have another member of the team look over it, as I'm not as familiar with the MQTT5 C++ bindings, but from everything I saw, it looks good to me!

File: source/mqtt/Mqtt5ClientCore.cpp [hunk 1]
```diff
@@ -0,0 +1,615 @@
...
+                AWS_ASSERT(callbackData->clientCore != nullptr);
+
+                /* callback not set */
+                if (callbackData->onSubscribeCompletion == NULL)
+                {
+                    goto on_subscribeCompletionCleanup;
+                }
+
+                {
+                    std::lock_guard<std::recursive_mutex> lock(callbackData->clientCore->m_callback_lock);
```
[the remark this answers is not in the document]
Comment (xiazhvera) [reply source/mqtt/Mqtt5ClientCore.cpp:328]: No worry, the lock will destroy itself when it goes out of scope. And the lock will get unlock at that time.  Here is an example: https://en.cppreference.com/w/cpp/thread/recursive_mutex

File: source/mqtt/Mqtt5ClientCore.cpp [hunk 2]
```diff
@@ -0,0 +1,615 @@
...
+
+                clientOptions.client_termination_handler = &Mqtt5ClientCore::s_clientTerminationCompletion;
+                clientOptions.client_termination_handler_user_data = this;
+
+                m_client = aws_mqtt5_client_new(allocator, &clientOptions);
+            }
+
+            Mqtt5ClientCore::~Mqtt5ClientCore()
+            {
+                // release lock
```
[the remark this answers is not in the document]
Comment (xiazhvera) [reply source/mqtt/Mqtt5ClientCore.cpp:468]: Sorry, I should remove this. This was a TODO reminder for myself to make sure I release the mutex and other resources. While after that, it turns out that we do not need to do anything to release the mutex on destroy.

File: source/mqtt/Mqtt5ClientCore.cpp [hunk 3]
```diff
@@ -0,0 +1,615 @@
...
+                options.completion_callback = Mqtt5ClientCore::s_publishCompletionCallback;
+                options.completion_user_data = pubCallbackData;
+
+                int result = aws_mqtt5_client_publish(m_client, &publish, &options);
+                if (result != AWS_OP_SUCCESS)
+                {
+                    Crt::Delete(pubCallbackData, pubCallbackData->allocator);
+                    return false;
+                }
+                return true;
```
[the remark this answers is not in the document]
Comment (xiazhvera) [reply source/mqtt/Mqtt5ClientCore.cpp:532]: Great catch!! Thanks!!

File: include/aws/crt/mqtt/Mqtt5Client.h [hunk 4]
```diff
@@ -146,58 +147,49 @@ namespace Aws
...
              * Type signature of the callback invoked when client connection stopped
              * Mandatory event fields: client
              */
-            using OnStoppedHandler = std::function<void(Mqtt5Client &, const OnStoppedEventData &)>;
-
-            /**
-             * Type signature of the callback invoked when a Disconnection Comlete
-             *
-             */
-            using OnDisconnectCompletionHandler = std::function<void(std::shared_ptr<Mqtt5Client>, int)>;
```
[the remark this answers is not in the document]
Comment (xiazhvera) [reply include/aws/crt/mqtt/Mqtt5Client.h:177]: Yes, the disconnect completion is not exposed in the C++. The cleanup process uses the client termination callback.

**Comment by xiazhvera** [comment]
Thanks for the review! As the PR introduce API changes, I will merge it with canary changes and other related PR changes together.

**Pull request merged**